state_test.go

```diff
@@ -0,0 +1,48 @@
+package frankenphp
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test2GoroutinesYieldToEachOtherViaStates(t *testing.T) {
+	threadState := &threadState{currentState: stateBooting}
+
+	go func() {
+		threadState.waitFor(stateInactive)
+		assert.True(t, threadState.is(stateInactive))
+		threadState.set(stateReady)
+	}()
+
+	threadState.set(stateInactive)
+	threadState.waitFor(stateReady)
+	assert.True(t, threadState.is(stateReady))
+}
+
+func TestStateShouldHaveCorrectAmountOfSubscribers(t *testing.T) {
+	threadState := &threadState{currentState: stateBooting}
+
+	// 3 subscribers waiting for different states
+	go threadState.waitFor(stateInactive)
+	go threadState.waitFor(stateInactive, stateShuttingDown)
+	go threadState.waitFor(stateShuttingDown)
+
+	time.Sleep(1 * time.Millisecond)
```

Comment:
This is flaky, you should use a WaitGroup instead (or check the number of subscribers in a loop)

Reply:
Yeah you're right, I'll change it to check in a loop.